### DISCUSSION TITLE
A pattern for i18n

### DISCUSSION THREAD

**Discussion by osdiab** [solidjs/solid-start#570] [Ideas]
Requesting that in some future version of Solid Start, there is a clear path to achieve i18n, ideally in a type-safe way (typed keys). If anything exists in that direction, resources would be appreciated - thanks for the great project!

**Comment by CanRau** [comment]
Want to add some references

[Typesafe i18n](https://github.com/ivanhofer/typesafe-i18n) - A fully type-safe and lightweight internationalization library for all your TypeScript and JavaScript projects.
Already provides an [adapter-solid](https://github.com/ivanhofer/typesafe-i18n/tree/main/packages/adapter-solid)

[SvelteKit PR for translatable/alternate route paths](https://github.com/sveltejs/kit/pull/1130#pullrequestreview-1217710621)

**Reply by nirtamir2** [reply]
Also - if you want to keep the "json" way to do it - we have tools to help stricten the types
I think i18next can support solid-js and it provides some type-safety out of the box.
`@solid-primitives/i18n` can be used with eslint plugins to verify existence - like https://github.com/nirtamir2/eslint-plugin-i18n-prefix and   https://github.com/natura-cosmeticos/eslint-plugin-i18n-checker

**Comment by BierDav** [comment]
I would really like to see a solution with ssr support. That's all we need 🙂

**Reply by atk** [reply]
Typesafe-i18n already supports SSR.

**Reply by BierDav** [reply]
Ahh really. I've already tried the provided example there, but this really doesn't support ssr. Do you have an example project how to implement, becuase I didn't get it to work.

**Reply by atk** [reply]
If I understand correctly, what's missing is a way for the i18n to be initialized from the request instead of client access.

**Reply by BierDav** [reply]
Yeah I guess, so my goal is to have ssr for all first page loads and then transform into a spa with only client side rendering. This also opens me the possibility to ssg all pages that don't require user authentication.
